mtp_api/apps/transaction/managers.py

```diff
@@ -1,25 +1,39 @@
-from datetime import timedelta
-
 from django.conf import settings
+from django.db import connection
 from django.db import models
 from django.db.transaction import atomic
 
+from credit.models import Credit
+from .constants import TRANSACTION_STATUS
+
 
 class TransactionQuerySet(models.QuerySet):
 
     @atomic
     def reconcile(self, date, user):
         update_set = self.filter(
-            received_at__gte=date,
-            received_at__lt=(date + timedelta(days=1)),
+            self.model.STATUS_LOOKUP[TRANSACTION_STATUS.RECONCILABLE],
+            received_at__date=date.date(),
             credit__isnull=False,
             credit__reconciled=False
         ).order_by('id').select_for_update()
 
+        ref_codes = []
         ref_code = settings.REF_CODE_BASE
         for transaction in update_set:
-            if transaction.reconcilable:
-                transaction.ref_code = ref_code
-                ref_code += 1
-            transaction.credit.reconcile(user)
-            transaction.save()
+            ref_codes.append((transaction.id, ref_code))
+            ref_code += 1
+
+        if ref_codes:
+            with connection.cursor() as c:
+                c.execute('DROP TABLE IF EXISTS refids;')
+                c.execute('CREATE TEMP TABLE refids (id integer, ref_code integer)')
+                insert_query = 'INSERT INTO refids (id, ref_code) VALUES '
+                insert_query += ', '.join(['%s' for _ in ref_codes])
+                insert_query += ';'
+                c.execute(insert_query, ref_codes)
+                c.execute('UPDATE transaction_transaction t SET '
+                          'ref_code=r.ref_code FROM refids r WHERE '
+                          't.id=r.id;')
```

Comment:
All of this raw SQL (and particularly the `join` in the `insert_query`) worries me. Is there anything in the ORM we can use instead of hand-crafting these?